### PR TITLE
Custom made Enumerable methods

### PR DESCRIPTION
**Pull request by ClintonEnyinna**
Nine custom made enumerable methods to practice the use of yield and Proc

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice Job :+1: 
I like the effort you put into this project. I tested your methods and they pass most edge cases however, we need to fix some failing ones before we can merge this PR. Check my comments for suggestions.

### Martin Afani

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,122 @@
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    each do |i|
+      yield i
+    end
+    self
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [x] identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)). returns an Enumerator if no block is given

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+  def my_each_with_index
+    (0...to_a.length).each do |i|
+      yield(to_a[i], i)
+    end
+    self
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [x] identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)). returns an enumerator if no block is given. You can achieve this by adding `return to_enum`

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+  def my_select
+    arr = []
+    to_a.my_each do |num|
+      yield num ? arr.push(num) : ''
+    end
+    arr
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [x] should be identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)). returns an array containing all elements of enum for which the given block returns a true value.
- [x] returns an enumerator if no block is given

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+  def my_all?
+    to_a.my_each do |num|
+      return false unless yield num
+    end
+    true
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [x] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when a class is passed as an argument returns true if all of the collection is a member of such class
- [x] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x] when no block or argument is given returns true when none of the collection members are false or nil
- [x] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+  def my_any?
+    to_a.my_each do |num|
+      return true if yield num
+    end
+    false
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [x] should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when no block or argument is given returns true if at least one of the collection is not false or nil
- [x] when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [x] when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+  def my_none?
+    if block_given?
+      to_a.my_any?(&Proc.new)
+    else
+      to_a.length.positive?
+    end
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [x] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] returns true if the block never returns true for all elements
- [x] when no block or argument is given returns true only if none of the collection members is true
- [x] when a class is passed as an argument returns true if none of the collection is a member of such class
- [x] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: enumerable.rb [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    count
+  end
+
+  def my_map
+    arr = []
+    to_a.my_each do |item|
+      arr.push(yield item)
+    end
+    arr
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [x] should be identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] returns an Enumerator if no block is given

File: enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+  def my_inject(arg = false)
+    if arg
+      acc = arg
+      (0...to_a.length).my_each { |i| acc = yield(acc, to_a[i]) }
+    else
+      acc = to_a[0]
+      (1...to_a.length).my_each { |i| acc = yield(acc, to_a[i]) }
+    end
+    acc
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a symbol is specified combines each element of the collection by applying the symbol as a named method

**Comment by shshamim63** [comment]
## Status :red_circle:  [!! Required Changes !!] :red_circle: 
___
Hi @ClintonEnyinna :wave: ,
Awesome job. Can see a well descriptive readme and you have covered some task mentioned by the previous reviewer :clap: . Still there are some issues left behind which are blocking to approve this project. But you will cover them easily. So, please follow the instructions below :point_down: to complete this project.
___
### :spaghetti: [Mandatory Changes] :spaghetti: 
___
1. **my_each** identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)). So you have to cover the functionality of each by maybe a loop to access all the items of an array. Not by invoking each inside my_each.
https://github.com/ClintonEnyinna/enumerable_methods/blob/6653a37fce29b6637437ca450da4cfe36ea1b678/enumerable.rb#L4-L11
---
2. **my_each_with_index**  identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)) You have used each method, which you were suppose to generate. You can use your my_each instead of each.
https://github.com/ClintonEnyinna/enumerable_methods/blob/6653a37fce29b6637437ca450da4cfe36ea1b678/enumerable.rb#L13-L20
---
3. [**my_all?**](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern. Suppose if `[3, 3, 3].my_all?(3)` get executed, should return true.
---
4. [**my_any?**](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
...1. Returns true if the block ever returns a value other than false or nil. Which means that if `[1, 2 , 3].my_any?` gets executed should return true. But your one returns false.
...2. When a class is passed as an argument returns true if at least one of the collection is a member of such class. For an example `[1, 2, 3].my_any?(Integer)` should return true.
In your code you have checked for `if arg`. So if it gets an arg then it will never gonna check for the if, elsif. 
...3. When a Regex is passed as an argument returns false if none of the collection matches the Regex. `%w[ cat bat cup ].my_any?(/t/)` should return true.
...4. When a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern `[1, 2, 3].my_any?(2)` should return true.
https://github.com/ClintonEnyinna/enumerable_methods/blob/6653a37fce29b6637437ca450da4cfe36ea1b678/enumerable.rb#L46
---
5. [**my_none?**](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)) Is totally opposite to my_any?
..1. Returns true if the block never returns true for all elements. `[1, 2 , 3].my_none?` gets executed should return false.
...2. When a class is passed as an argument returns true if none of the collection is a member of such class. Has the same problem mentioned above on my_any no.2.
...3. When a Regex is passed as an argument returns true only if none of the collection matches the Regex. `%w[ cat bat cup ].my_none?(/t/)` should return false.
...4. When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern. `[1, 2, 3].my_none?(2)` should return false.
---
6. [**my_inject**](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
It can take upto two parameter(init, symbol). So if `[1, 2, 3].my_inject(2, :+)` gets executed should return 8.
___
#### Please fix the above mentioned issue and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice Job :+1: 
You have done well implementing the suggested changes. However, one method is still failing.
Check my comment and re-send for review after making the suggested changes.

### Martin Afani

File: enumerable.rb [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+      else
+        acc = args
+        (0...to_a.length).my_each { |i| acc = yield(acc, to_a[i]) }
+      end
+    else
+      acc = to_a[0]
+      (1...to_a.length).my_each { |i| acc = yield(acc, to_a[i]) }
+    end
+    acc
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
This is what I get after testing that method:
inject

Comment (ClintonEnyinna) [reply enumerable.rb]: How do I fix this? It returns the same thing as Ruby's inject

Comment (whiz25) [reply enumerable.rb]: @ClintonEnyinna 
You can do something like this;
`def my_inject(arg_1 = nil, arg_2 = nil, &block)
    array = to_a.dup
    (inject, sym, array) = get_block_and_symbol(arg_1, arg_2, array, block_given?)
    inject = if sym
               symbol(array, sym, inject)
             else
               block(array, inject, &block)
             end
    inject
  end`

Helper methods;
`def get_block_and_symbol(arg1, arg2, arr, block)
    arg1 = arr.shift if arg1.nil? && block
    return [arg1, nil, arr] if block
    return [arr.shift, arg1, arr] if arg2.nil?
    [arg1, arg2, arr]
  end`

`def symbol(array, sym, inject)
    array[0..-1].my_each { |i| inject = inject.send(sym, i) }
    inject
  end`

`def block(array, inject, &block)
    raise LocalJumpError, 'no block given' unless block
    array[0..-1].my_each { |i| inject = block.yield(inject, i) }
    inject
  end`

That should fix the problem.

**Review by whiz25 — APPROVED** [review]
## Awesome job :clap: 
It's a :+1: from me. Merge and proceed.

## Martin Afani

**Pull request closed without merging**